### DISCUSSION TITLE
Is there any way to subset_fonts before inserting fonts?

### DISCUSSION THREAD

**Discussion by izerui** [pymupdf/PyMuPDF#3383] [Looking for help] [answered]
I can know all the text before I do the page insertion, this will improve the efficiency of subset_fonts, it seems like subset_fonts doesn't support multithreading.
Creating subsets is a time-consuming action when working with documents with many pages

**Comment by JorjMcKie** [comment, marked as answer]
Threading is not supported at all in PyMuPDF! Please use the `multiprocessing` module of Python when needed.

Creating font subsets is now directly supported by MuPDF. It still is declared as "experimental", so we haven't yet included it officially.
But you can use it and we do not know of any issues. It is at least 15 times faster than the current method.
Here is what you can do instead `doc.subset_fonts()`:
```python
pdf = fitz._as_pdf_document(doc)  # access underlying PDF-specific level
fitz.mupdf.pdf_subset_fonts2(pdf, list(range(doc.page_count)))
```

**Reply by izerui** [reply]
Thanks!!!, you provide the compression method code block, I tried, it seems that the size of the file is not reduced, I'm not quite sure what is the reason for this,  I am using page.insert_htmlbox way to add the text.

**Reply by izerui** [reply]
Sorry, I wrote a separate test case and it seems like he took effect, I'll check my program again.

**Reply by JorjMcKie** [reply]
you have of course to use a save method with all available compression!!!
`doc.ez_save(...)`.

**Reply by izerui** [reply]
Yes, just now I also found out that if I use save then the size of the save is the same, if it is ez_save then the file size will be much smaller, by the way I would like to ask again: is ez_save the same processing logic as doc.tobytes(garbage=4, deflate=True), thank you very much!

**Reply by JorjMcKie** [reply]
`doc.tobytes(garbage=4, deflate=True, use_objstms=1)`.
Object stream support (use_objstms) provides for additional compression effects.

**Reply by izerui** [reply]
Your tips, and improve the compression effect of my final pdf, very ideal, thank you so much, obviously my negligence, not timely attention to the source code parameters of some of the description.

**Comment by izerui** [comment]
Can I use this code block in multiple threads? 
This allows multiple documents to be compressed at the same time!

```
pdf = fitz._as_pdf_document(doc)  # access underlying PDF-specific level
fitz.mupdf.pdf_subset_fonts2(pdf, list(range(doc.page_count)))
```

**Reply by JorjMcKie** [reply]
**_NO! No multi-threading support!_**
Put processing different PDF's in different processes using Python `multiprocessing` module.
And do not forget to use `doc.ez_save()`!

**Reply by izerui** [reply]
Thanks, I see what you mean, I use different threads in the thread pool to compress different documents, and I haven't found any conflicting issues for the time being, can I test it for a while and see if it exposes some issues?
```
docs = [doc0, doc1, doc2....]
with concurrent.futures.ThreadPoolExecutor(max_workers=10) as pool:
        
        pdf = fitz._as_pdf_document(docs[i])  # access underlying PDF-specific level
          fitz.mupdf.pdf_subset_fonts2(pdf, list(range(docs[i].page_count)))

          tmp_dir = tempfile.gettempdir()
          tmp_file_path = os.path.join(tmp_dir, f"{tmp_dir}/{int(time.perf_counter() * 1000)}.pdf")
          doc.ez_save(tmp_file_path)
          doc.close()
          return fitz.open(tmp_file_path)

```

Forgive me for wanting to try for faster performance.

**Reply by JorjMcKie** [reply]
You can do what you want of course.

I would definitely use multiprocessing instead. It is safer **_and_** also faster than multithreading, because it is the operating system doing the parallelism - not Python. There is an example in the documentation.
